crates/uv-distribution-filename/src/expanded_tags.rs

```diff
@@ -0,0 +1,459 @@
+/// ```
+///
+/// This type stores those expanded tags.
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct ExpandedTags(smallvec::SmallVec<[WheelTag; 1]>);
+
+impl ExpandedTags {
+    /// Parse a list of expanded wheel tags (e.g., `py3-none-any`).
+    pub fn parse<'a>(tags: impl IntoIterator<Item = &'a str>) -> Result<Self, ExpandedTagError> {
+        let tags = tags
+            .into_iter()
+            .map(parse_expanded_tag)
+            .collect::<Result<_, _>>()?;
+        Ok(Self(tags))
+    }
+
+    /// Returns `true` if the wheel is compatible with the given tags.
+    pub fn is_compatible(&self, compatible_tags: &Tags) -> bool {
+        self.0.iter().any(|tag| {
+            compatible_tags.is_compatible(tag.python_tags(), tag.abi_tags(), tag.platform_tags())
+        })
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum ExpandedTagError {
+    #[error("The wheel tag \"{0}\" is missing a language tag")]
+    MissingLanguageTag(String),
+    #[error("The wheel tag \"{0}\" is missing an ABI tag")]
+    MissingAbiTag(String),
+    #[error("The wheel tag \"{0}\" is missing a platform tag")]
+    MissingPlatformTag(String),
+    #[error("The wheel tag \"{0}\" contains too many segments")]
+    ExtraSegment(String),
+    #[error("The wheel tag \"{0}\" contains an invalid language tag")]
+    InvalidLanguageTag(String, #[source] ParseLanguageTagError),
+    #[error("The wheel tag \"{0}\" contains an invalid ABI tag")]
+    InvalidAbiTag(String, #[source] ParseAbiTagError),
+    #[error("The wheel tag \"{0}\" contains an invalid platform tag")]
+    InvalidPlatformTag(String, #[source] ParsePlatformTagError),
+}
+
+/// Parse an expanded (i.e., simplified) wheel tag, e.g. `py3-none-any`.
+///
+/// Unlike parsing tags in a wheel filename, each tag in this case is expected to contain exactly
+/// three segments separated by `-`: a language tag, an ABI tag, and a platform tag; however,
+/// empirically, some build backends do emit multipart tags (like `cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64`),
+/// so we allow those too.
+fn parse_expanded_tag(tag: &str) -> Result<WheelTag, ExpandedTagError> {
+    let mut splitter = memchr::Memchr::new(b'-', tag.as_bytes());
+    if tag.is_empty() {
+        return Err(ExpandedTagError::MissingLanguageTag(tag.to_string()));
+    }
+    let Some(python_tag_index) = splitter.next() else {
+        return Err(ExpandedTagError::MissingAbiTag(tag.to_string()));
+    };
+    let Some(abi_tag_index) = splitter.next() else {
+        return Err(ExpandedTagError::MissingPlatformTag(tag.to_string()));
+    };
+    if splitter.next().is_some() {
+        return Err(ExpandedTagError::ExtraSegment(tag.to_string()));
+    }
+
+    let python_tag = &tag[..python_tag_index];
+    let abi_tag = &tag[python_tag_index + 1..abi_tag_index];
+    let platform_tag = &tag[abi_tag_index + 1..];
+
+    let is_small = memchr(b'.', tag.as_bytes()).is_none();
+
+    if let Some(small) = is_small
```

Comment:
Here, we're ignoring `Invalid*Tag` error types, while filtering out tags we don't recognise. The same problem exists in the original wheel filename implementation where `InvalidLanguageTag` is unused (`InvalidLanguageTag` has two hits in the codebase, this file and `wheel.rs`). In both cases we have `repr: SmallString` to cover tags we filtered out in error messages.

Reply:
Yeah this is intentional, because we don't want to error out on tags we don't understand.